Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_self_intersections.h

```diff
@@ -759,86 +943,270 @@ bool construct_tentative_hole_patch(std::vector<typename boost::graph_traits<Tri
+  return true;
+}
 
-      // ignore boundary edges
-      if(vi+1 == vj || (vj == 0 && static_cast<std::size_t>(vi) == cc_border_vertices.size()-1))
-        continue;
+template <typename GeomTraits>
+struct Mesh_projection_functor
+{
+  typedef typename GeomTraits::Point_3 Point_3;
+  typedef typename GeomTraits::Triangle_3 Triangle_3;
+
+  typedef std::vector<Triangle_3> Triangle_container;
+  typedef CGAL::AABB_triangle_primitive<GeomTraits, typename Triangle_container::const_iterator> Primitive;
+  typedef CGAL::AABB_traits<GeomTraits, Primitive> Traits;
+  typedef CGAL::AABB_tree<Traits> Tree;
+
+  template <typename TriangleMesh, typename VPM>
+  Mesh_projection_functor(const TriangleMesh& mesh,
+                          const VPM vpm)
+  {
+    triangles.reserve(num_faces(mesh));
+    for(auto f : faces(mesh))
+      triangles.emplace_back(get(vpm, target(halfedge(f, mesh), mesh)),
+                             get(vpm, target(next(halfedge(f, mesh), mesh), mesh)),
+                             get(vpm, source(halfedge(f, mesh), mesh)));
+
+    tree.insert(std::cbegin(triangles), std::cend(triangles));
+  }
+
+  Point_3 operator()(const Point_3& p) const { return tree.closest_point(p); }
+
+private:
+  Triangle_container triangles;
+  Tree tree;
+};
+
+// Really rough and absurdly high complexity (at least factorize the AABB tree...)
+template <typename Point, typename Projector, typename TriangleMesh, typename GeomTraits>
+bool adapt_patch(std::vector<std::vector<Point> >& point_patch,
+                 const Projector& projector,
+                 const TriangleMesh& tmesh,
+                 const GeomTraits& gt)
+{
+  typedef typename boost::graph_traits<TriangleMesh>::vertex_descriptor        vertex_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::halfedge_descriptor      halfedge_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::edge_descriptor          edge_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::face_descriptor          face_descriptor;
 
-      halfedge_descriptor h = halfedge(cc_border_vertices[vi], cc_border_vertices[vj], tmesh).first;
-      if(h != boost::graph_traits<TriangleMesh>::null_halfedge() &&
-         cc_interior_edges.count(edge(h, tmesh)) == 0)
+  typedef typename GeomTraits::FT FT;
+
+  std::vector<Point> soup_points;
+  std::vector<std::array<FT, 3> > soup_faces;
+
+  FT avg_edge_length = 0;
+  int pid = 0;
+  std::map<Point, std::size_t> point_ids;
+  for(const auto& fp : point_patch)
+  {
+    CGAL_assertion(fp.size() == 3);
+    std::array<FT, 3> f;
+    for(std::size_t i=0; i<3; ++i)
+    {
+      avg_edge_length += CGAL::approximate_sqrt(squared_distance(fp[i], fp[(i+1)%3]));
+      auto res = point_ids.emplace(fp[i], pid);
+      if(res.second)
       {
-        non_manifold_edge_found = true;
-        break;
+        soup_points.push_back(fp[i]);
+        ++pid;
       }
+      f[i] = res.first->second;
```

Comment:
`warning C4244: '=': conversion from 'std::size_t' to 'double', possible loss of data`